src/main/utility/mod.rs

```diff
@@ -362,70 +383,112 @@ impl std::fmt::Display for VerifyPluginPathError {
                 f.write_str("permission denied traversing path")
             }
             VerifyPluginPathError::UnhandledIoError(e) => write!(f, "unhandled io error: {e}"),
+            VerifyPluginPathError::IncompatibleInterpreter(e) => {
+                write!(f, "script with incompatible interpreter: {e}")
+            }
+            VerifyPluginPathError::UnknownFileType => f.write_str("Uncrecognized file type"),
         }
     }
 }
 
 /// Check that the plugin path is executable under Shadow.
-pub fn verify_plugin_path(path: impl AsRef<std::path::Path>) -> Result<(), VerifyPluginPathError> {
-    let path = path.as_ref();
-
-    let metadata = std::fs::metadata(path).map_err(|e| {
-        match e.kind() {
-            std::io::ErrorKind::NotFound => VerifyPluginPathError::NotFound,
-            std::io::ErrorKind::PermissionDenied => VerifyPluginPathError::PathPermissionDenied,
-            // TODO handle TooManyLinks when stabilized
-            // TODO handle InvalidFileName when stabilized
-            k => {
-                log::warn!("Unhandled error getting metadata for {path:?}: {k:?}");
-                VerifyPluginPathError::UnhandledIoError(e)
-            }
-        }
-    })?;
-
+fn verify_plugin_path_internal(
+    path: impl AsRef<std::path::Path> + std::fmt::Debug,
+) -> Result<(), VerifyPluginPathError> {
+    let file = std::fs::File::open(&path)?;
+    let metadata = file.metadata()?;
     if !metadata.is_file() {
         return Err(VerifyPluginPathError::NotFile);
     }
-
     // this mask doesn't guarantee that we can execute the file (the file might have S_IXUSR
     // but be owned by a different user), but it should catch most errors
     let mask = libc::S_IXUSR | libc::S_IXGRP | libc::S_IXOTH;
     if (metadata.mode() & mask) == 0 {
+        log::debug!("{path:?} not executable");
         return Err(VerifyPluginPathError::NotExecutable);
     }
+    // Get up to PATH_MAX bytes; that should be enough to ensure we get the
+    // interpreter name where applicable.
+    let mut reader = std::io::BufReader::with_capacity(linux_api::limits::PATH_MAX, file);
+    let buf = reader.fill_buf()?;
```

Comment:
I think `fill_buf` only makes a single `read` syscall and doesn't loop, so it won't necessarily read as many bytes as we want. It might be better to do `file.take(PATH_MAX).read_to_end()` instead?

Reply:
Thanks; yeah that's better